### PR TITLE
Add new filters for individual sitemap entries

### LINKED ISSUE #188: Add a filter to get_url_list() in class Core_Sitemaps_Posts
@swissspidy 
### Description
 A plugin author cannot add any additional POST attributes to the url_list array at the individual POST (or POST ID) level within the For Loop. 

**Which feature is your enhancement request related to?**
https://github.com/GoogleChromeLabs/wp-sitemaps/pull/183

**Describe the solution you'd like**
Allow plugin developers to filter the url_list[] array at the post level within the  For Loop in the function get_url_list() of Class  Core_Sitemaps_Posts.  A filter "**core_sitemaps_post_attributes**" within the POSTs  loop will be better for a developer to add those additional attributes. Currently, the function adds the post attribute "loc" for URL to the array. Making a filter for the array as shown below will let plugin authors to add their post attributes to the array. This is how we see it, 

`foreach ( $posts as $post ) {

 $post_sitemap_attributes = array( 'loc' => get_permalink( $post )) ;

 $url_list[] = apply_filters( 'core_sitemaps_post_attributes', $post_sitemap_attributes, $post );

}`

**Additional context**
I have included a file from a plugin that adds the additional POST attributes using this proposed filter.
[class-core-sitemaps-posts.txt](https://github.com/GoogleChromeLabs/wp-sitemaps/files/4672986/class-core-sitemaps-posts.txt)

### Acceptance Criteria
Add the relevant acceptance criteria here.
- [My code follows WordPress coding standards.]
- [I have performed a self-review of my own code.]
- [My changes generate no new warnings.]

### LINKED ISSUE #189: Add a filter to get_url_list() in class Core_Sitemaps_Taxonomies
@swissspidy 
### Description
 A plugin author cannot add any additional Term attributes to the url_list array at the individual Taxonomy Term (or TERM ID) level within the For Each Loop. 

**Which feature is your enhancement request related to?**
https://github.com/GoogleChromeLabs/wp-sitemaps/pull/183

**Describe the solution you'd like**
Allow plugin developers to filter the url_list[] array at the Term level within the  Taxonomy Terms ForEach Loop in the function get_url_list() of Class  Core_Sitemaps_Taxonomies.  A filter "**core_sitemaps_term_attributes**" within the Taxonomy Terms ForEach loop will be better for a developer to add those additional attributes. Currently, the function adds the post attribute "loc" for Term Link to the array. Making a filter for the array as shown below will let plugin authors to add their Term attributes to the array. This is how I see it, 

`
              if ( ! empty( $taxonomy_terms->terms ) ) {
	            foreach ( $taxonomy_terms->terms as $term ) {
			    
	                $term_sitemap_attributes = array( 'loc' => get_term_link( $term )) ;                   
                       $url_list[] = apply_filters( 'core_sitemaps_term_attributes', $term_sitemap_attributes, $term );        	        
       
	          }
           }`

**Additional context**
I have included a file from a plugin that adds the additional Term attributes using this proposed filter.

[class-core-sitemaps-taxonomies.txt](https://github.com/GoogleChromeLabs/wp-sitemaps/files/4673000/class-core-sitemaps-taxonomies.txt)

### Acceptance Criteria
Add the relevant acceptance criteria here.
- [My code follows WordPress coding standards.]
- [I have performed a self-review of my own code.]
- [My changes generate no new warnings.]

### PR DESCRIPTION
**Pull request by swissspidy**
### Issue Number

Fixes #188.
Fixes #189.

### Description

Adds new filters to more easily add custom attributes for individual items, providing post/term/user objects to the consumer.

### Screenshots (before and after if applicable)
N/A

### Type of change
Please select the relevant options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Enhancement (change which improves an existing feature. E.g., performance improvement, docs update, etc.)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Steps to test
Describe the tests required to verify your changes.
Provide instructions so the PR Tester can check functionality and also list any relevant details and / or dependencies required for your tests.

### Acceptance criteria
- [x] My code follows WordPress coding standards.
- [x] I have performed a self-review of my own code.
- [ ] If the changes are visual, I have cross browser / device tested.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added test instructions that prove my fix is effective or that my feature works.

### CODE REVIEW

File: inc/providers/class-core-sitemaps-posts.php [hunk 1]
```diff
@@ -103,9 +103,21 @@ public function get_url_list( $page_num, $post_type = '' ) {
...
 			);
+
+			/**
+			 * Filters the sitemap entry for an individual post.
+			 *
+			 * @since 5.5.0
+			 *
+			 * @param array   $sitemap_entry Sitemap entry for the post.
+			 * @param WP_Post $post Post object.
+			 * @param string  $post_type Name of the post_type.
```
Comment (felixarntz) [inline inc/providers/class-core-sitemaps-posts.php]: Descriptions should be indented.

File: inc/providers/class-core-sitemaps-taxonomies.php [hunk 2]
```diff
@@ -98,9 +98,21 @@ public function get_url_list( $page_num, $taxonomy = '' ) {
...
 				);
+
+				/**
+				 * Filters the sitemap entry for an individual term.
+				 *
+				 * @since 5.5.0
+				 *
+				 * @param array   $sitemap_entry Sitemap entry for the term.
+				 * @param WP_Term $term Term object.
+				 * @param string  $taxonomy Taxonomy name.
```
Comment (felixarntz) [inline inc/providers/class-core-sitemaps-taxonomies.php]: See above.

File: inc/providers/class-core-sitemaps-users.php [hunk 3]
```diff
@@ -42,9 +42,20 @@ public function get_url_list( $page_num, $object_subtype = '' ) {
...
 				'loc'     => get_author_posts_url( $user->ID ),
 			);
+
+			/**
+			 * Filters the sitemap entry for an individual user.
+			 *
+			 * @since 5.5.0
+			 *
+			 * @param array   $sitemap_entry Sitemap entry for the user.
+			 * @param WP_User $user User object.
```
Comment (felixarntz) [inline inc/providers/class-core-sitemaps-users.php]: See above.

**Review by felixarntz — COMMENTED** [review]
+1 for these filters. Should we maybe remove the URL list filters? They seem a bit redundant at this point:
* The new `*_entry` filters here allow modifying an entry's data.
* The query filters from #182 allow modifying which entries are included in a sitemap (modifying the query is anyway a cleaner approach I'd argue than "randomly" removing individual items after querying which would also result in inconsistent page counts).

FWIW I think the current `*_url_list` filters are fully covered by both of the above filter groups in a much better way, using the `*_url_list` filters is more complicated regardless of what of the above two things you're trying to achieve.

**Comment by swissspidy** [comment]
> Should we maybe remove the URL list filters?

Yeah I think so 👍

**Comment by swissspidy** [comment]
> Should we maybe remove the URL list filters?

I'll do that in a follow-up PR once we #183 lands, as it also  requires changes to the docs.

**Comment by swissspidy** [comment]
Opened #195

**Review by felixarntz — APPROVED** [review]

**Comment by Rajesh-Techblissonline** [comment]
@swissspidy @felixarntz  These filter need to be supported by a filter that lets third party developers to create and present their own sitemap-xml using their custom stylesheets. These custom stylesheets may vary, depending on the object type (or object sub-type). Currently, I am doing it in my plugin as explained in Issue #151 [here](https://github.com/GoogleChromeLabs/wp-sitemaps/issues/151#issuecomment-633191537).  Pls. share your thoughts. You may check the working  example sitemaps - [here ](https://techblissonline.com/sitemap.xml)

**Pull request merged**